Neurodamus bump and implementing stable dependency

ferdonline opened the pull request:
Creates version specification which depend on both the model and core versions.

E.g. using model 1.1 and core 3.0.1 it will define a version '1.1-3.0.1' which uses tag 1.1 and depends on core 3.0.1

ferdonline committed:
Neurodamus bump and implementing stable dependency

iomaganaris commented on var/spack/repos/builtin/packages/py-neurodamus/package.py:
```diff
@@ -13,6 +13,7 @@ class PyNeurodamus(PythonPackage):
     git      = "ssh://bbpcode.epfl.ch/sim/neurodamus-py"
 
     version('develop', branch='master')
+    version('2.1.0',   tag='2.1.0')
```
Because of some changes in this version with the interface to `CoreNEURON` there is incompatibilities with the previous versions of `neurodamus-core`. Maybe we should do something similar to the versions dependencies with `py-neurodamus` as well?

ferdonline replied:
Neurodamus-py doesn't depend on neurodamus-core/model..., because it expects a mechlib (from any model) at runtime. But we should have runtime checks, early if possible.

What wouldn't work in previous neurodamus-py? 
We should be careful with providing compat with (maybe, at least) the previous version. What do you think?

iomaganaris replied:
I changed this here https://bbpcode.epfl.ch/code/#/c/51154/3/hoc/Node.hoc and here https://bbpcode.epfl.ch/code/#/c/51155/8/neurodamus/node.py. So previous hoc versions get different arguments. Can we somehow check which neurodamus-core version we have then from neurodamus-py?

ferdonline commented:
Ohh.. damn pep8! But CI tests pass :)

ferdonline committed:
Pep8 fixes

matz-e commented on deploy/packages/bbp-packages.yaml:
```diff
@@ -168,11 +168,12 @@ packages:
-      - neurodamus-neocortex+coreneuron@1.0^coreneuron+knl
-      - neurodamus-hippocampus+coreneuron@1.0^coreneuron+knl
-      - neurodamus-thalamus+coreneuron@1.0^coreneuron+knl
+      # Can we deploy always the latest? Versions are now bit more stable (and long)
```
With the current implementation, this will generate modules for all neurodamus versions found.

matz-e replied:
The next iteration of the deployment would solve this…

ferdonline replied:
last commit adds `@1.1-3.0.2` but the deployments scripts don't like it

matz-e replied:
I'm not surprised...

ferdonline replied:
Actually it was due to the double `@@` typo :)
In the end, if you don't think is too bizarre I would like to go forward with it, and no worries, you can blame me for any problem, I will do my best at maintaining it. Just le me know if you see a better way at achieving a similar result, or you are happy with the implementation

matz-e replied:
It still looks freakish to me, personally. Remove that comment above, and I'll approve, though. Technically nothing wrong with it…

matz-e commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -14,6 +14,18 @@
+    for model_v, core_v in model_core_dep_v:
+        this_version = model_v + "-" + core_v  # e.g. 1.1-3.0.2
+        version(this_version, tag=model_v, submodules=True, get_full_repo=True)
+        depends_on("neurodamus-core@" + core_v, when="@" + this_version)
```
I don't know if I agree with this yet.

matz-e commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -14,6 +14,18 @@
+def version_from_model_core_deps(model_core_dep_v):
+    """Creates version specification which depend on both the model
+       and core versions.
+       E.g. using model 1.0 and core 3.0.1 it will define a version
```
Should this be 1.0 → 1.1?

ferdonline replied:
sure, it was just an example

pramodk commented:
If I am not mistaken this would be first package in Spack to do dynamic version generation and hence I am hesitant about this.  Technically seems ok but is this way we should adopt?

Apart from neurodamus use case, this issue of specific version requirement exist everywhere. I am wondering if we should solve this issue in neurodamus in different way e.g. using `resource`.

I will let @matz-e  to comment on this better...

ferdonline commented:
> dynamic version generation

We still statically maintain the list of versions, it's not automatic

> Apart from neurodamus use case, this issue of specific version requirement exist everywhere. I am wondering if we should solve this issue in neurodamus in different way e.g. using `resource`.

Don't agree. Neurodamus-model is a special case because it *incorporates* code from the core (making 90% of the code or so). So actually every time we bumped the core, neurodamus-xxx changed and the version didn't. It has been a long pending issue to have neurodamus-xxx versions being meaningful. I don't think this logic should be used anywhere else.

ferdonline commented:
> dynamic version generation

We still statically maintain the list of versions, it's not automatic

pramodk commented:
> Don't agree. Neurodamus-model is a special case because it incorporates code from the core

My comment is in the context of packages like LLVM where you can see what happens: 
https://github.com/BlueBrain/spack/blob/967c78dfb9ecaa4eb3a080836f6198a66d88af6e/var/spack/repos/builtin/packages/llvm/package.py#L99

LLVM `composed` of 10 different components. The question is how you define the version of LLVM if you bump up the version of one of it's `resource`.

ferdonline committed:
Take neurodamus-py 2.1.1. Deploy spec with version

ferdonline committed:
Merge branch 'develop' into neurodamus/bump_improv_dep

ferdonline commented:
As far as I see in LLVM, all those extra components follow the same version. 
I don't know of many cases where we actually put two independent codes, supposedly developed by different people, following completely different versions and make a single thing. I also think that knowing each individual version is very helpful for both developer and users. 
That was the original intent. This is not an attempt to solve the spack multi-deployments-per-version problem, although it helps.
Anyway, I don't claim this is a perfect solution, and if you believe this would be useful for other packages, please go ahead and improve it to make more generic. I just feel happy with the result given the effort.

matz-e commented on deploy/packages/bbp-packages.yaml:
```diff
@@ -41,7 +41,7 @@ packages:
       - py-morph-tool@0.2.5
       - py-morphio@2.3.4
-      - py-neurodamus
+      - py-neurodamus@@2.1.1
```
```suggestion
      - py-neurodamus@2.1.1
```

ferdonline committed:
Update deploy/packages/bbp-packages.yaml

Co-authored-by: Matthias Wolf <m+git@sushinara.net>

ferdonline committed:
Addressing review comments

ferdonline committed:
Neurodamus-model core dependency set to build-only

Previously the dependency was set to 'run' as well to force updating
the package whenever core was updated. This no longer applies since
the core dependency version is now explicit.

ferdonline committed:
Merge branch 'develop' into neurodamus/bump_improv_dep

pramodk commented:
> I don't know of many cases where we actually put two independent codes, supposedly developed by different people, following completely different versions and make a single thing. I also think that knowing each individual version is very helpful for both developer and users.

this kind of done via submodule right? e.g. coreneuron is incomplete without mod2c. So we have mod2c as submodule into coreneuron (and mod2c can be deployed as separate standalone version).

Technically this change looks Ok. Just that more logic into package recipe caught my attention. 

By the way, looking at [module usage on BB5](https://bbpmonitoring.epfl.ch/metrics/d/OojPuR5Wz/bb5-module-usage-on-bb5?orgId=1&from=now-90d&to=now), I am wondering if / how people are using neurodamus-core.

ferdonline commented:
please retest

ferdonline commented:
> this kind of done via submodule right? e.g. coreneuron is incomplete without mod2c. 

@pramodk maybe we can have a call if you want to discuss this and look for solution together. I would say it's still a different case, since both are perfectly independent applications which, only for user convenience, we put together in a pack.  Let me elaborate my view better.
Since the split of neurodamus from the models the two things became effectively independent. We can have only the core and we can have only the model, the advantage being that scientists can easily develop and compile their models and integrate with one neurodamus-core.
The way this integration is done can nowadays be twofold: (1) the old way of "compile together" and (2) since January, compile special from model only and load libs, which is what bglib-py does it. (For the record is also supported in neurodamus-py: load `neurodamus-core~commonmods`, set `BGLIBPY_MOD_LIBRARY_PATH` and launch)
So, IMO, this is a software pack "neurodamus+model" (hence the name). However we were using the model version to version the "pack", which is IMO bad. Because of that, even when we added new important features to core, core+model didn't see a new version, which probably confused users. We saw that most of them even started relying on the Archives, since the version was meaningful-less: neurodamus-neocortex@0.4 from May was a very different thing from neurodamus@0.4 from October.
I have acknowledged this problem since long but the only solution I saw before was to have Git projects only with two submodules: core and model... but that insanely complicates the setup.
Given this python implementation ended being relatively simple I thought we can use it in this rare case. But if you have ideas let me know, I would be happy to solve the problem in a better way
